storage/mysql/mysql_test.go

```diff
@@ -28,14 +67,335 @@ func TestMain(m *testing.M) {
+
+	db, err := sql.Open("mysql", *mysqlURI+"?multiStatements=true")
+	if err != nil {
+		klog.Fatalf("Failed to connect to DB: %v", err)
+	}
+	defer func() {
+		if err := db.Close(); err != nil {
+			klog.Warningf("Failed to close db: %v", err)
+		}
+	}()
+
+	if _, err := db.ExecContext(ctx, dropTablesSQL); err != nil {
+		klog.Fatalf("Failed to drop all tables: %v", err)
+	}
+
+	if _, err := db.ExecContext(ctx, string(rawSchema)); err != nil {
+		klog.Fatalf("Failed to execute init database schema: %v", err)
+	}
+}
+
+func TestNew(t *testing.T) {
+	ctx := context.Background()
+
+	for _, test := range []struct {
+		name string
+		opts []func(*tessera.StorageOptions)
+	}{
+		{
+			name: "no tessera.StorageOption",
+			opts: nil,
+		},
+		{
+			name: "standard tessera.WithCheckpointSignerVerifier",
+			opts: []func(*tessera.StorageOptions){
+				tessera.WithCheckpointSignerVerifier(noteSigner, noteVerifier),
+			},
+		},
+		{
+			name: "all tessera.StorageOption",
+			opts: []func(*tessera.StorageOptions){
+				tessera.WithCheckpointSignerVerifier(noteSigner, noteVerifier),
+				tessera.WithBatching(1, 1*time.Second),
+				tessera.WithPushback(10),
+			},
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			if _, err := mysql.New(ctx, testDB, test.opts...); err != nil {
+				t.Errorf("got err: %v", err)
+			}
+		})
+	}
+}
+
+func TestReadMissingTile(t *testing.T) {
+	ctx := context.Background()
+	s := newTestMySQLStorage(t, ctx)
+
+	for _, test := range []struct {
+		name                string
+		level, index, width uint64
+	}{
+		{
+			name:  "0/0/0",
+			level: 0, index: 0, width: 0,
+		},
+		{
+			name:  "123/456/789",
+			level: 123, index: 456, width: 789,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			tile, err := s.ReadTile(ctx, test.level, test.index, test.width)
+			if err != nil {
+				t.Errorf("got err: %v", err)
+			}
+			if tile != nil {
+				t.Error("tile is not nil")
+			}
+		})
+	}
+}
+
+func TestReadMissingEntryBundle(t *testing.T) {
+	ctx := context.Background()
+	s := newTestMySQLStorage(t, ctx)
+
+	for _, test := range []struct {
+		name  string
+		index uint64
+	}{
+		{
+			name:  "0",
+			index: 0,
+		},
+		{
+			name:  "123456789",
+			index: 123456789,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			entryBundle, err := s.ReadEntryBundle(ctx, test.index)
+			if err != nil {
+				t.Errorf("got err: %v", err)
+			}
+			if entryBundle != nil {
+				t.Error("entryBundle is not nil")
+			}
+		})
+	}
+}
+
+func TestAdd(t *testing.T) {
+	ctx := context.Background()
+	s := newTestMySQLStorage(t, ctx)
+
+	for _, test := range []struct {
+		name      string
+		entry     []byte
+		wantIndex uint64
+	}{
+		{
+			name:      "empty string entry",
+			entry:     []byte(""),
+			wantIndex: 0,
+		},
+		{
+			name:      "123 string entry",
+			entry:     []byte("123"),
+			wantIndex: 1,
+		},
+		{
+			name:      "empty byte",
+			entry:     []byte{},
+			wantIndex: 2,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
```

Comment:
These test case functionally depend on each other for the `wantIndex` value, so I suggest that we remove the pretense that they are separate test cases? It is idiomatic in my experience with go that table-driven tests each start with a clean environment.

Reply:
Gotcha, I've removed `TestAdd`. `TestParallelAdd` has the test coverage for `mysql.Stoage.Add`.